servers/physics_3d/shape_3d_sw.cpp

```diff
@@ -1640,8 +1654,177 @@ Vector3 HeightMapShape3DSW::get_support(const Vector3 &p_normal) const {
 	return get_aabb().get_support(p_normal);
 }
 
+void HeightMapShape3DSW::_face_cull_segment(_SegmentCullParams &p_params) const {
+	Vector3 res;
+	Vector3 normal;
+	if (p_params.face->intersect_segment(p_params.from, p_params.to, res, normal)) {
+		real_t d = p_params.dir.dot(res) - p_params.dir.dot(p_params.from);
+		if ((d > 0) && (d < p_params.min_d)) {
+			p_params.min_d = d;
+			p_params.result = res;
+			p_params.normal = p_params.face->normal;
+			p_params.normal = normal;
+			p_params.collisions++;
+		}
+	}
+}
+
+void HeightMapShape3DSW::_cell_cull_segment(_SegmentCullParams &p_params, int p_x, int p_z) const {
+	// First triangle.
+	_get_point(p_x, p_z, p_params.face->vertex[0]);
+	_get_point(p_x + 1, p_z, p_params.face->vertex[1]);
+	_get_point(p_x, p_z + 1, p_params.face->vertex[2]);
+	p_params.face->normal = Plane(p_params.face->vertex[0], p_params.face->vertex[1], p_params.face->vertex[2]).normal;
+	_face_cull_segment(p_params);
+
+	// Second triangle.
+	_get_point(p_x + 1, p_z, p_params.face->vertex[0]);
+	_get_point(p_x + 1, p_z + 1, p_params.face->vertex[1]);
+	p_params.face->normal = Plane(p_params.face->vertex[0], p_params.face->vertex[1], p_params.face->vertex[2]).normal;
+	_face_cull_segment(p_params);
+}
+
 bool HeightMapShape3DSW::intersect_segment(const Vector3 &p_begin, const Vector3 &p_end, Vector3 &r_point, Vector3 &r_normal) const {
-	return false;
+	if (heights.is_empty()) {
+		return false;
+	}
+
+	Vector3 local_begin = p_begin + local_origin;
+	Vector3 local_end = p_end + local_origin;
+
+	FaceShape3DSW face;
+	face.backface_collision = false;
+
+	_SegmentCullParams params;
+	params.from = p_begin;
+	params.to = p_end;
+	params.dir = (p_end - p_begin).normalized();
+	params.face = &face;
+
+	// Quantize the ray begin/end.
+	int begin_x = floor(local_begin.x);
+	int begin_z = floor(local_begin.z);
+	int end_x = floor(local_end.x);
+	int end_z = floor(local_end.z);
+
+	if ((begin_x == end_x) && (begin_z == end_z)) {
+		// Simple case for rays that don't traverse the grid horizontally.
+		// Just perform a test on the given cell.
+		int x = CLAMP(begin_x, 0, width - 2);
+		int z = CLAMP(begin_z, 0, depth - 2);
```

Comment:
I didn't realize that, it can be dangerous in some cases indeed. For this case it seems like it wouldn't make a huge difference anyway so I can probably leave it this way for readability.